### ISSUE TITLE
Getting topics and probabilities from loaded model

### ISSUE THREAD

**Issue by ginward** [MaartenGr/BERTopic#283]
I understand that, if you are fitting the model freshly, you can get the topic and probablities of each document by doing the following: 

`topics, probs = topic_model.fit_transform(docs)
`

However, if you are reading the topics from a file, how do you get topics and probablities from it? 

`topic_model = BERTopic.load("bert_model")`

Is there a method to get the topics and their probabilties for each document?

**Comment by ginward** [comment]
@MaartenGr  Is it possible to just call `transform` on `topic_model`?

**Comment by ginward** [comment]
I am currently setting `calculate_probabilities=False`. But will the above method you proposed set it to `True`?

**Comment by ginward** [comment]
Yes, I am. Thanks!

**Comment by antoninabondarchuk** [comment]
Hi @MaartenGr !

I used these suggested lines of code after loading my model:
`topics = topic_model._map_predictions(topic_model.hdbscan_model.labels_)

probs = topic_model.hdbscan_model.probabilities_`

But I realized that it returned only 165k of documents instead of 270k it was trained on. 
Why could it be the case? 

Thanks in advance for your time!

**Comment by MaartenGr** [comment]
@antoninabondarchuk That is indeed rather strange! Instead of going through what is happening here, it might be worthwhile to use the newly released v0.12 version of BERTopic. It has fixed some of the issues and difficulties with having to use `._map_predictions` in order to extract the topics. Instead, you can now do the following:

```python
topics = topic_model.topics_
probabilities = topic_model.probabilities_
```

**Comment by rja122277** [comment]
I'm using Bertopic: 1.1.2, but the codes don't work. It says,"'BERTopic' object has no attribute 'topics_'". Could you help me with solving this problem?

My ultimate goal is to get topic-probabilities matrix for each document. I find that topic_model.visualize_distribution(probs[n]) produces the information exactly I want, but I don't know how to get their information in the form I can export to csv file. Is the approach suggested in this thread the best way get topics-probs matrix for all documents?  Thank you!

**Comment by rja122277** [comment]
I'm using Bertopic: 1.1.2, but the codes don't work. It says,"'BERTopic' object has no attribute 'topics_'". Could you help me with solving this problem?

My ultimate goal is to get topic-probabilities matrix for each document. I find that topic_model.visualize_distribution(probs[n]) produces the information exactly I want, but I don't know how to get their information in the form I can export to csv file. Is the approach suggested in this thread the best way get topics-probs matrix for all documents? Thank you!

**Comment by MaartenGr** [comment]
@rja122277 

> I'm using Bertopic: 1.1.2, but the codes don't work. It says,"'BERTopic' object has no attribute 'topics_'". Could you help me with solving this problem?

There is currently no v1.1.2. However, there is a v0.12 version that should support the `.topics_` attribute. Did you try upgrading to the latest version? 

> My ultimate goal is to get topic-probabilities matrix for each document. I find that topic_model.visualize_distribution(probs[n]) produces the information exactly I want, but I don't know how to get their information in the form I can export to csv file. Is the approach suggested in this thread the best way get topics-probs matrix for all documents? Thank you!

With v0.12, you should be able to perform the following:

```python
topics = topic_model.topics_
probabilities = topic_model.probabilities_
```

That way, you can export the `probabilities` variable to a csv using pandas. Just make sure you have set `calculate_probabilities=True` in BERTopic.

**Comment by brandywwy** [comment]
Hi @MaartenGr , thank you so much for your great work! 
I have been trying to save and load model to present. However, when I finish loading the model and try to get flat probabilities, it returns BaseCluster instead of my origin HDBSCAN model, below is my code:

```
# cuML version HDBSCAN
hdbscan_model = HDBSCAN(min_cluster_size=300, min_samples=50, cluster_selection_method='eom', verbose=6)
vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words="english", min_df = 0.1, max_df=0.7)
representation_model = KeyBERTInspired()

topic_model = BERTopic(nr_topics='auto',
                          hdbscan_model=hdbscan_model,
                          vectorizer_model=vectorizer_model,
                          representation_model=representation_model,
                          seed_topic_list=seed_topic_list,
                          verbose=True)
```


After doing .fit_Transform() and get the results, I save the model:

```
# Save model
embedding_model = "all-MiniLM-L6-v2"
topic_model.save("test_model", serialization="pytorch", save_ctfidf=True, save_embedding_model=embedding_model)
```

Then load it:
```
# load model
loaded_model = BERTopic.load("test_model")
```

And then I got the error below:
![Screenshot 2023-08-16 at 12 43 46 PM](https://github.com/MaartenGr/BERTopic/assets/43160664/14449591-7140-4bcf-9bdc-21eb9462197e)


Could you please help to see where I am doing wrong? Any advice would be highly appreciated! Thank you.

**Comment by MaartenGr** [comment]
@brandywwy Thank you for your kind words!

When you save the model using either `pytorch` or `safetensors`, the underlying HDBSCAN and UMAP models are not saved. Instead, inference will be done by comparing the document embeddings with the topic embeddings. This results in a much smaller model (often times at least 10x less the size), faster inference, and safer serialization. You can find more about serialization [here](https://maartengr.github.io/BERTopic/getting_started/serialization/serialization.html).

**Comment by z3ugma** [comment]
I do want to use the vectors produced by topic probabilities as suggested in `all_points_membership_vectors` to store in a vector DB and power a vector search. Is that a good approach?

**Comment by MaartenGr** [comment]
@z3ugma Sure, that is definitely a decent approach. Note that this is essentially what `.transform` is doing if you save and load the model using either safetensors or pytorch.

**Comment by rayenFathallah** [comment]
If I want to get for a document the different probabilities for each topic, In which format should i export the model ?

**Comment by MaartenGr** [comment]
@rayenFathallah That doesn't matter. Any option will give you back the probabilities. Having said that, pickle will save the underlying dim & cluster models whilst the other methods do not. Which both have its own (dis)advantages.

**Comment by MohamedRebai41** [comment]
Hi @MaartenGr , thank you so much for your amazing work!

I serialized my model to `safetensors` format. When I load it, I can't seem to find a way to retrieve or recalculate the probabilities of each document to each topic. Since the `hdbscan` and `umap` models are not saved with the model. I can't use the soft clustering provided by the `hdbscan` library. Is there any other way of getting probabilty by topic for each document?

**Comment by rayenFathallah** [comment]
@MohamedRebai41 I did the same thing, what I want to achieve is that I want to calculate the top 10 topics for a document. when I transform it It only returns the first topic. Even if in the training I used calculate_probabilities = True. This occurs when extracting the model using safetensors. So I think we should extract the entire model using pytroch in order to predict the probabilities for each topic given a document. @MaartenGr  Is that correct ?

**Comment by MaartenGr** [comment]
@MohamedRebai41 You still can calculate the probabilities using `.transform` if you have set `calculate_probabilities=True`. Even though `hdbscan` isn't used, we calculate it using the cosine similarity between topic and document embeddings.
